Adds ESS imaging workflow

dtasev opened the pull request:
This PR adds the ESS imaging workflow converted to use `scipp` for the reduction of the data. It follows the steps and uses the parameters from this [ESS reduction script](https://git.esss.dk/testbeamline/gp2/blob/master/reduce.py).

It also adds a notebook containing timings with the goal of benchmarking different steps. That notebook should not be used for reduction as it does not have all the steps performed on all the data.

Commit:
Adds ESS imaging workflow

OwenArnold commented on scipp-imaging-2D-Bench.ipynb:
```diff
@@ -0,0 +1,1585 @@
+    "pulse_nr_plastic = 1.0 / 2614343\n",
+    "\n",
+    "# let's get the process started:\n",
+    "experiment_dir = \"/home/dtasev/dev/scipp_imaging/RAL_Mantid_Sep2018/\"\n",
```
No hard-coded absolute paths please.

dtasev replied:
I'm not sure how to handle this. Absolute paths aren't good, but having the relative one isn't useful either, when this notebook isn't part of the `gp2` repo. The only way it can work is to be dependent on the `gp2` repository for the `config.py` (which will be done via a

```python
from utilities import get_this_dir

```
import). Is that acceptable?

OwenArnold replied:
I don't have to be dependent on that repository, but I think the same approach could be used. The `make_config` + products could be local to this repo.

OwenArnold commented on scipp-imaging-2D-Bench.ipynb:
```diff
@@ -0,0 +1,1585 @@
+    "\n",
+    "# These use the range notation of [start, stop, step]\n",
+    "# for R825\n",
+    "frame_parameters = [(15450,22942,64),\n",
```
Could you add more information here. This is more public facing that the scripts you originally worked from. What are these values, what units. Maybe could immediately be encoded as scipp variables as a self-documenting approach?

OwenArnold commented on scipp-imaging-2D.ipynb:
```diff
@@ -0,0 +1,8350 @@
+    "import os\n",
+    "\n",
+    "# Specify where the data can be found\n",
+    "experiment_dir = \"/home/dtasev/dev/scipp_imaging/RAL_Mantid_Sep2018/\"\n",
```
No absolute paths please!

OwenArnold commented on scipp-imaging-2D.ipynb:
```diff
@@ -0,0 +1,8350 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# created by Owen Arnold (optional e-mail) on 2018-09-19\n",
```
This is not correct, but you might reference the original repository and then set yourself as the author of the present script.

OwenArnold commented on scipp-imaging-2D.ipynb:
```diff
@@ -0,0 +1,8350 @@
+    "nx_target = grouping_number\n",
+    "ny_target = grouping_number\n",
+    "\n",
+    "frame_parameters = [(15167,23563,64),\n",
```
Could you add more information here. This is more public facing that the scripts you originally worked from. What are these values, what units. Maybe could immediately be encoded as scipp variables as a self-documenting approach?

OwenArnold commented on scipp-imaging-2D.ipynb:
```diff
@@ -0,0 +1,8350 @@
+    "\n",
+    "    return frames[0]\n",
+    "\n",
+    "def remove_special_values(values):\n",
```
Shouldn't this be `zero_special_values`?

dtasev replied:
I have renamed it as that name does make more sense.

OwenArnold commented on scipp-imaging-2D.ipynb:
```diff
@@ -0,0 +1,8350 @@
+      "GP2_007-003-WFM-open-beam_0853.tiff: Image 854, of 2001\r",
+      "GP2_007-003-WFM-open-beam_0854.tiff: Image 855, of 2001\r",
+      "GP2_007-003-WFM-open-beam_0855.tiff: Image 856, of 2001\r",
+      "GP2_007
```
x_start, x_end can be moved to outer block.

dtasev replied:
Good catch!

OwenArnold commented on scipp-imaging-2D.ipynb:
```diff
@@ -0,0 +1,8350 @@
+      "GP2_007-003-WFM-open-beam_0853.tiff: Image 854, of 2001\r",
+      "GP2_007-003-WFM-open-beam_0854.tiff: Image 855, of 2001\r",
+      "GP2_007-003-WFM-open-beam_0855.tiff: Image 856, of 2001\r",
+      "GP2_007
```
Question - given that the normalisation is detector by detector, is it more efficient to normalise then group or the other way around?

dtasev replied:
It is, but it changes the result values significantly from Mantid's (as division becomes the last step). 

At this point I think it's more important they stayed the same, at least until I do the value comparisons.

OwenArnold commented on scipp-imaging-2D.ipynb:
```diff
@@ -0,0 +1,8350 @@
+      "GP2_007-003-WFM-open-beam_0853.tiff: Image 854, of 2001\r",
+      "GP2_007-003-WFM-open-beam_0854.tiff: Image 855, of 2001\r",
+      "GP2_007-003-WFM-open-beam_0855.tiff: Image 856, of 2001\r",
+      "GP2_007
```
Was it a conscious decision not to use the functions from `compat/mantid`?

dtasev replied:
Good catch, `load_component_info_from_instrument_file` didn't exist (although the others did but I copied them for ease). But it is in the scipp PR now so I'll remove this block

Commit:
Use local config.py for paths. Adds more comments for parameters

Adds `make_config.py` to generate a `config.py` containing the data path.
Adds it to the README, and a sanity check in the notebook.

A .gitignore is added for the config.py and other commonly ignored files.

Cleans the output in both imaging notebooks.

Commit:
Now uses mantid compat fit

OwenArnold commented on scipp-imaging-2D.ipynb:
```diff
@@ -0,0 +1,496 @@
+    "    Loads all tiff images from the directory into a scipp Variable.\n",
+    "    \"\"\"\n",
+    "    stack = _load_tiffs(tiff_dir)\n",
+    "    return sc.Variable([Dim.Tof, Dim.Spectrum], \n",
```
Have you tried `float32`? Maybe more relevant for the benchmark results, but selecting precision is one of the reasons for working with scipp?

dtasev replied:
I have now, and I've added a new table for float32 [here](https://github.com/scipp/scipp/issues/812). It essentially doubles the speed in the slower parts, but the quick ones remain the same or very close

OwenArnold commented on scipp-imaging-2D.ipynb:
```diff
@@ -0,0 +1,496 @@
+   "outputs": [],
+   "source": [
+    "# If you want to save the workspace uncomment this line\n",
+    "SaveNexus(ws, \"scipp_normalized_wl.nxs\")"
```
`SaveNexusESS` should probably be used. Note that your `to_workspace_2d` would need to generate an instrument and attach it to output from scipp to make this fully useful.

dtasev replied:
I'll add the instrument as a change for https://github.com/scipp/scipp/pull/830

OwenArnold replied:
Its unfortunate that we didn't get round to bypassing instrument 1.0 in the instrument 2.0 construction otherwise you could have created the instrument directly from your arrays.

OwenArnold commented on scipp-imaging-2D.ipynb:
```diff
@@ -0,0 +1,496 @@
+    "# which is spectrum as outer-most dimension.\n",
+    "y = np.transpose(stitched[\"normalized_wl\"].values)\n",
+    "\n",
+    "ws = sc.compat.mantid.to_workspace_2d(x, y, None, x_dim)\n",
```
Presumably the `None` here is for the `error`. Can't we propagate errors in this script? We could assign uncertainties at the point of tiff loading.

dtasev replied:
Would they be just np.sqrt of the y values?

OwenArnold replied:
standard deviation is sqrt counts, but variance is sq of that. So the counts again.

Commit:
Use SaveNexusESS

Commit:
Loads instrument file into workspace before fit. Adds variances during loading

Commit:
Add DTYPE as parameter for the notebook

dtasev commented:
I think this is ready for another review @OwenArnold

OwenArnold commented on scipp-imaging-2D.ipynb:
```diff
@@ -0,0 +1,498 @@
+    "# How to start\n",
+    "\n",
+    "Before starting you must:\n",
+    "- Ensure that `scipp` and `mantid` are on your `PYTHONPATH`.\n",
```
These nice instructions use of config etc are not mirrored in the benchmark ipynb

dtasev replied:
Good catch. I've updated it to point to the non-bench notebook for instructions as I don't want to duplicate those, because they will get out of date. I've also added a link to the scipp issue with previous benchmarks.

Commit:
Update instructions and link benchmark issue
